### ISSUE TITLE
Comparing traces by latency

### ISSUE THREAD

**Issue by yurishkuro** [jaegertracing/jaeger-ui#513]
@tiffon you previously showed this diff mode at KubeCon'18, if I recall. What's the status of this, is the code somewhere in a branch?  It would be great to merge it.

![image](https://user-images.githubusercontent.com/3523016/73217801-9f931280-4126-11ea-9d88-532a0a53c39e.png)

(Asked in internal SO post 43701)

**Comment by yurishkuro** [comment]
Perhaps this is the branch: https://github.com/jaegertracing/jaeger-ui/tree/diff-on-latency

**Comment by tiffon** [comment]
@yurishkuro yes, that's the branch. A lot has changed, but I'll check it out this weekend.

**Comment by yurishkuro** [comment]
Thanks, Joe. This feels like it should be relatively small change on top of the existing diffs, but people do want it, even as experimental.

**Comment by tiffon** [comment]
@yurishkuro I didn't get a chance to work on the trace durations this weekend. I will try to take a look at it this week.

The old code is ~650 commits behind master and dates back to when we used flow. It was also a prototype for troubleshooting issues during a data-center turn-up. So, realistically, merging it is isn't going to work; it needs to be implemented from scratch.

I've also been thinking it would be nice to be able to represent the impact of nodes in the DAG that aren't common between the two traces. For instance, all of the common nodes could have nearly the same durations and the difference in trace durations is accounted for by nodes only in one or the other trace. It might be possible to just compare the average duration of the spans that are present against 0. That would only work for non-relative comparisons but relative comparisons were too noisy, anyway. Either way, it will be easier to explore this once something is up and running, but I wanted to mention it in case you have any suggestions.

**Comment by yurishkuro** [comment]
iirc, the graph already shows missing/added nodes in black & white (as opposed to red/green for comparable nodes).

**Comment by everett980** [comment]
@tiffon Idr the formula for creating dense nodes out of multiple spans, but if spans `a:0, a:1, b:0` are present in trace A three times, and present in trace B four times and result in one node (`data.a.members.length === 9` and `data.b.members.length === 12`) what durations would be compared?
Is it total duration of `data.a.members` vs total duration of `data.b.members`?
Does it make sense to instead/toggle compare the average duration?
What about max duration of `a:0, a:1, b:0` for each? This probably requires a different schema than what is in [PR#521](https://github.com/jaegertracing/jaeger-ui/pull/521)

**Comment by yurishkuro** [comment]
I think we shouldn’t be making any assumptions. One option is to not aggregate sibling paths at all, thus removing the problem. If user chooses to aggregate identical sibling paths, then they should specify the aggregation function to use with comparison, from min, max, avg.

**Comment by tiffon** [comment]
@everett980 the formula was average duration, and the only way data.a.members.length would be nine is if a:0 a:1 and b:0 were all grouped in the same node in the diff. 

@yurishkuro it's fine to change that, but I think it should be a separate ticket. This ticket and the PR that lays the ground work for implementing it should keep the grouping consistent with structural trace diffs and trace graphs. Or, if we want to prioritize changes to the structure / grouping of spans then we should add latency comparisons after that (IMO). I completely agree that would be a useful change (we already have a ticket for it) but I don't think we should combine it with duration diffs.

Also, grouping related siblings has upsides in terms of simplifications, mainly regarding ordering (IIRC) bc order then matters if siblings aren't grouped. And, with that, comparing  traces requires heuristics for matching, which is a high cost for the benefit of not grouping. Not intractable,  but not free either.

**Comment by everett980** [comment]
> the formula was average duration

I think there is value in having a toggle for total duration comparisons. If the number of grouped nodes in traceB has increased but the average time is unchanged (or even slightly reduced) then it should optionally reflect an increase in time spent on that node.
Or, due to potential overlap in a grouping (and this relates to the [Trace Statistics view](https://github.com/jaegertracing/jaeger-ui/pull/506) as well as the Trace Graph view) I think it would be beneficial to have a concept of "non-concurrent time" and "non-concurrent self time" when presenting a group of spans. 20 concurrent nodes with a (self) time of 30ms each is very different than 20 sequential nodes with (self) time of 30ms.

> and all either have children or are leaves.

For both the existing functionality and the duration functionality it would be nice to toggle this distinction (i.e.: DDG's "Show paths to the focal node" vs "Separate internal from external nodes"), but as mentioned this could be part of a separate issue.
